File: laundry-reservation/typescript/tests/RecordingMachineDevice.ts
import type { MachineDevice } from '../src/MachineDevice.js';

export interface LockCall {
  reservationId: string;
  reservationDateTime: Date;
  pin: number;
}

export class RecordingMachineDevice implements MachineDevice {
  readonly lockCalls: LockCall[] = [];
  readonly unlockCalls: string[] = [];
  shouldAcceptLock = true;

  lock(reservationId: string, reservationDateTime: Date, pin: number): boolean {
    this.lockCalls.push({ reservationId, reservationDateTime, pin });
    return this.shouldAcceptLock;
  }

  unlock(reservationId: string): void {
    this.unlockCalls.push(reservationId);
  }
}

File: laundry-reservation/python/tests/recording_machine_device.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime


@dataclass
class LockCall:
    reservation_id: str
    reservation_date_time: datetime
    pin: int


class RecordingMachineDevice:
    def __init__(self) -> None:
        self.lock_calls: list[LockCall] = []
        self.unlock_calls: list[str] = []
        self.should_accept_lock: bool = True

    def lock(self, reservation_id: str, reservation_date_time: datetime, pin: int) -> bool:
        self.lock_calls.append(LockCall(reservation_id, reservation_date_time, pin))
        return self.should_accept_lock

    def unlock(self, reservation_id: str) -> None:
        self.unlock_calls.append(reservation_id)
